2019-08-03-q-and-as-about-nix-for-haskellers.md

```diff
@@ -0,0 +1,692 @@
+    main documentation for the Nix language and CLI tools is in the
+    [Nix manual](https://nixos.org/nix/manual/).  The main documentation for the
+    derivations and helper functions in nixpkgs is the
+    [nixpkgs manual](https://nixos.org/nixpkgs/manual/).
+
+    For the most part, these manuals are written as reference manuals rather
+    than beginner-oriented HOW-TO manuals. As a beginner, it is quite difficult
+    to figure out how to do things just from the above manuals.[^4]
+
+2.  Nix is a dynamic language.  Unlike Haskell, you don't get "documentation for
+    free" with type signatures.
+
+    There are no tools similar to [Haddock](https://www.haskell.org/haddock/)
+    or [doctest](https://github.com/sol/doctest) for Nix.
+
+3.  nixpkgs is a lot of things to a lot of people.  There are a large number of
+    people using nixpkgs for widely different things.  There are people who use
+    it purely as a repository for system packages (like zlib, glibc, GTK, etc).
+    There are people who use it for easily generating a GHC package database
+    containing packages like aeson, lens, etc for Haskell development.  There
+    are people who use it for doing Python development.  There are people who
+    use it for cross compiling ARM binaries on x86.  And many more.
+
+    All of these groups of people have their own code in nixpkgs that others
+    often don't venture into.  For example, I've been using Nix for
+    about a year and a half, but I don't really have any idea how the
+    Python libraries are structured, how cross compiling works, etc.
+
+    If I wanted to play around with cross compiling, it would be time-consuming
+    to come up to speed with how everything works.
+
+Learning Nix is a non-trivial endeavour.  However, Nix is extremely flexible.
+Like Haskell, I'd say the time-investment can be worth it depending on how
+extensively you're planning on using it.
+
+With the above knowledge, it should be possible to dive into the questions from
+Dmitrii.
+
+## Questions and Answers
+
+Below, I list Dmitrii's questions and answer them one-by-one.[^7]  These answers
+should be understandable by people with no experience using Nix.
+
+### 1. What is the difference between `nix-build` and `nix-shell`?
+
+`nix-build` and `nix-shell` are the two most widely used Nix tools.
+
+They both evaluate derivations in a similar way, but their final results are
+different.
+
+When `nix-build` is finished evaluating a derivation, it produces a directory
+in the Nix store containing the binaries, shared objects, header files, etc for
+the derivation.
+
+When `nix-shell` is finished evaluating a derivation, it launches a Bash shell.
+It makes all the inputs to the derivation available in the Bash shell.  For
+instance, imagine you had a derivation that depends on `git`.  If you ran
+`nix-shell` on that derivation, it would put you in a Bash shell with an
+environment where the `git` CLI tools available.  From `nix-shell`,
```

Comment:
Maybe it will require to restructure blog post a little bit, but I think it would be super helpful to show in one place (somewhere on the first lines) the benefits of `nix-shell` for development. Like a numbered list with use cases. Like this:

1. Your package performs call to some CLI tool (like `git` or `hub`). So for this package to work you need to install this tool first. And if you don't install it, you package will crash, and probably in a bad way. Nix allows you to specify such dependencies automatically so you don't need to worry about installing anything manually.
2. To build some packages you need to have some C libraries (or system libraries for them). Again, nix allows you to specify them so you don't need to worry about having them in advance.

I think this really helps to understand the benefits of `nix`.